### PR TITLE
Add more types to Pyodide glue code

### PR DESCRIPTION
**Pull request by garrettgu10**
Quick wins woohoo

### CODE REVIEW

**Comment by garrettgu10** [comment]
Looks like there was an uncaught bug in setupPackages where an array was being coerced to a string. It should be fixed now, but that's a good argument for why this work is necessary.

File: src/pyodide/internal/jaeger.ts [hunk 1]
```diff
@@ -1,10 +1,9 @@
-// @ts-nocheck
 import { default as internalJaeger } from "pyodide-internal:internalJaeger";
 
 /**
  * Used for tracing via Jaeger.
  */
-export function enterJaegerSpan(span, callback) {
+export function enterJaegerSpan(span: String, callback: Function) {
```
Comment (hoodmane) [inline src/pyodide/internal/jaeger.ts:6]:
```suggestion
export function enterJaegerSpan<T>(span: string, callback: () => T): T {
```

File: src/pyodide/internal/loadPackage.ts [hunk 2]
```diff
@@ -17,11 +16,16 @@ import { parseTarInfo } from "pyodide-internal:tar";
...
 import { createTarFS } from "pyodide-internal:tarfs";
 
-async function decompressArrayBuffer(arrBuf) {
-  return await new Response(new Response(arrBuf).body.pipeThrough(new DecompressionStream("gzip"))).arrayBuffer();
+async function decompressArrayBuffer(arrBuf: ArrayBuffer): Promise<ArrayBuffer> {
+  const resp = new Response(arrBuf);
+  if (resp && resp.body) {
+    return await new Response(resp.body.pipeThrough(new DecompressionStream("gzip"))).arrayBuffer();
+  } else {
+    throw "Failed to decompress array buffer";
```
Comment (hoodmane) [inline src/pyodide/internal/loadPackage.ts:24]: Don't throw bare strings
```suggestion
    throw new Error("Failed to decompress array buffer");
```

File: src/pyodide/internal/jaeger.ts [hunk 3]
```diff
@@ -1,10 +1,9 @@
-// @ts-nocheck
 import { default as internalJaeger } from "pyodide-internal:internalJaeger";
 
 /**
  * Used for tracing via Jaeger.
  */
-export function enterJaegerSpan(span, callback) {
+export function enterJaegerSpan<T>(span: String, callback: () => T): T {
```
Comment (hoodmane) [inline src/pyodide/internal/jaeger.ts:6]: Lowercase s:
```suggestion
export function enterJaegerSpan<T>(span: string, callback: () => T): T {
```

File: src/pyodide/internal/setupPackages.ts [hunk 4]
```diff
@@ -67,9 +69,9 @@ class SitePackagesDir {
    * @param {List<String>} soFiles A list of .so files contained in the small bundle
    * @param {String} requirement The canonicalized package name this small bundle corresponds to
    */
-  addSmallBundle(tarInfo, soFiles, requirement) {
+  addSmallBundle(tarInfo: FSInfo, soFiles: string[], requirement: string) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:72]:
```suggestion
  addSmallBundle(tarInfo: FSInfo, soFiles: string[], requirement: string): void {
```

File: src/pyodide/internal/setupPackages.ts [hunk 5]
```diff
@@ -82,18 +84,18 @@ class SitePackagesDir {
    * @param {List<String>} soFiles A list of .so files contained in the big bundle
    * @param {List<String>} requirements canonicalized list of packages to pick from the big bundle
    */
-  addBigBundle(tarInfo, soFiles, requirements) {
+  addBigBundle(tarInfo: FSInfo, soFiles: string[], requirements: Set<string>) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:87]:
```suggestion
  addBigBundle(tarInfo: FSInfo, soFiles: string[], requirements: Set<string>): void {
```

File: src/pyodide/internal/setupPackages.ts [hunk 6]
```diff
@@ -135,7 +137,7 @@ export function buildSitePackages(requirements) {
  *
  * TODO: stop using loadPackage in workerd.
  */
-export function patchLoadPackage(pyodide) {
+export function patchLoadPackage(pyodide: {loadPackage: Function}) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:140]:
```suggestion
export function patchLoadPackage(pyodide: {loadPackage: Function}): void {
```

File: src/pyodide/internal/setupPackages.ts [hunk 7]
```diff
@@ -147,14 +149,14 @@ function disabledLoadPackage() {
...
-function getTransitiveRequirements() {
+function getTransitiveRequirements(): Set<string> {
   const requirements = REQUIREMENTS.map(canonicalizePackageName);
   // resolve transitive dependencies of requirements and if IN_WORKERD install them from the cdn.
   const packageDatas = recursiveDependencies(LOCKFILE, requirements);
   return new Set(packageDatas.map(({ name }) => canonicalizePackageName(name)));
 }
 
-export function getSitePackagesPath(Module) {
+export function getSitePackagesPath(Module: Module) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:159]:
```suggestion
export function getSitePackagesPath(Module: Module): string {
```

File: src/pyodide/internal/setupPackages.ts [hunk 8]
```diff
@@ -168,7 +170,7 @@ export function getSitePackagesPath(Module) {
  * details, so even though we want these directories to be on sys.path, we
  * handle that separately in adjustSysPath.
  */
-export function mountLib(Module, info) {
+export function mountLib(Module: Module, info: FSInfo) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:173]:
```suggestion
export function mountLib(Module: Module, info: FSInfo): void {
```

File: src/pyodide/internal/setupPackages.ts [hunk 9]
```diff
@@ -187,15 +189,15 @@ export function mountLib(Module, info) {
  * Add the directories created by mountLib to sys.path.
  * Has to run after the runtime is initialized.
  */
-export function adjustSysPath(Module) {
+export function adjustSysPath(Module: Module) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:192]:
```suggestion
export function adjustSysPath(Module: Module): void {
```

File: src/pyodide/internal/setupPackages.ts [hunk 10]
```diff
@@ -187,15 +189,15 @@ export function mountLib(Module, info) {
...
   const site_packages = getSitePackagesPath(Module);
   simpleRunPython(
     Module,
     `import sys; sys.path.append("/session/metadata"); sys.path.append("${site_packages}"); del sys`,
   );
 }
 
-function recursiveDependencies(lockfile, names) {
+function recursiveDependencies(lockfile: PackageLock, names: string[]) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:200]:
```suggestion
function recursiveDependencies(lockfile: PackageLock, names: string[]): string[] {
```

File: src/pyodide/internal/setupPackages.ts [hunk 11]
```diff
@@ -210,7 +212,7 @@ function recursiveDependencies(lockfile, names) {
  * @param toLoad The set of names of packages to load
  * @private
  */
-function addPackageToLoad(lockfile, name, toLoad) {
+function addPackageToLoad(lockfile: PackageLock, name: string, toLoad: Map<string, PackageDeclaration>) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:215]:
```suggestion
function addPackageToLoad(lockfile: PackageLock, name: string, toLoad: Map<string, PackageDeclaration>): void {
```

File: src/pyodide/types/runtime-generated/metadata.d.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+declare namespace MetadataReader {
+  const isWorkerd: () => boolean;
+  const isTracing: () => boolean;
+  const shouldSnapshotToDisk: () => boolean;
+  const isCreatingBaselineSnapshot: () => boolean;
+  const getRequirements: () => Array<string>;
```
Comment (hoodmane) [inline src/pyodide/types/runtime-generated/metadata.d.ts:6]:
```suggestion
  const getRequirements: () => string[];
```

File: src/pyodide/internal/setupPackages.ts [hunk 13]
```diff
@@ -14,12 +13,12 @@ const canonicalizeNameRegex = /[-_.]+/g;
...
  * @private
  */
-function canonicalizePackageName(name) {
+function canonicalizePackageName(name: string): string {
   return name.replace(canonicalizeNameRegex, "-").toLowerCase();
 }
 
 // The "name" field in the lockfile is not canonicalized
-const STDLIB_PACKAGES = Object.values(LOCKFILE.packages)
+const STDLIB_PACKAGES: string[] = Object.values(LOCKFILE.packages)
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:21]: You'll want to add a file `pyodide-internal:generated/pyodide-lock.json.d.ts` to give the types for `LOCKFILE`.

**Comment by hoodmane** [comment]
Can you format everything by running `npx prettier -w .` in the `src/pyodide` folder?

File: src/pyodide/internal/setupPackages.ts (hunk 13 above)
Comment (garrettgu10) [reply src/pyodide/internal/setupPackages.ts:21]: Right, I had that, and I forgot to git add it 😪

File: src/pyodide/internal/builtin_wrappers.ts [hunk 14]
```diff
@@ -105,7 +105,10 @@ function prepareStackTrace(_error: Error, stack: StackItem[]) {
...
 }
 
-export async function wasmInstantiate(module: WebAssembly.Module | Uint8Array, imports: WebAssembly.Imports) {
+export async function wasmInstantiate(
+  module: WebAssembly.Module | Uint8Array,
+  imports: WebAssembly.Imports,
+) {
   if (!(module instanceof WebAssembly.Module)) {
     checkCallee();
     module = UnsafeEval.newWasmModule(module);
```
Comment (hoodmane) [inline src/pyodide/internal/builtin_wrappers.ts:114]:
```suggestion
export async function wasmInstantiate(
  mod: WebAssembly.Module | Uint8Array,
  imports: WebAssembly.Imports,
): { module: WebAssembly.Module, instance: WebAssembly.Instance } {
  let module;
  if (mod instanceof WebAssembly.Module)) {
    module = mod;
  } else {
    checkCallee();
    module = UnsafeEval.newWasmModule(module);
```

File: src/pyodide/internal/setupPackages.ts [hunk 15]
```diff
@@ -135,26 +147,28 @@ export function buildSitePackages(requirements) {
  *
  * TODO: stop using loadPackage in workerd.
  */
-export function patchLoadPackage(pyodide) {
+export function patchLoadPackage(pyodide: { loadPackage: Function }) {
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:150]:
```suggestion
export function patchLoadPackage(pyodide: { loadPackage: Function }): void {
```

**Review by hoodmane — APPROVED** [review]
Thanks @garrettgu10!

**Comment by hoodmane** [comment]
Tests seem to be failing though.

File: src/pyodide/internal/loadPackage.ts [hunk 16]
```diff
@@ -12,16 +11,29 @@
...
+  arrBuf: ArrayBuffer,
+): Promise<ArrayBuffer> {
+  const resp = new Response(arrBuf);
+  if (resp && resp.body) {
+    return await new Response(
+      resp.body.pipeThrough(new DecompressionStream("gzip")),
+    ).arrayBuffer();
+  } else {
+    throw new Error("Failed to decompress array buffer");
+  }
```
Comment (hoodmane) [inline src/pyodide/internal/loadPackage.ts:33]: I think the following code would be slightly cleaner here (though it's pretty unimportant):
```js
const blob = new Blob([arrBuf]);
const decompressed = blob.stream().pipeThrough(new DecompressionStream("gzip"));
return await new Response(decompressed).arrayBuffer();
```

**Pull request merged**